### PR TITLE
Adding support for reading crt/key from Vault

### PR DESCRIPTION
**Pull request by ekingme**
Adding support for reading crt/key from Vault

### CODE REVIEW

File: spec/defines/vhost_spec.rb [hunk 1]
```diff
@@ -57,4 +60,22 @@
...
+
+  context 'with vault => true' do
+    let(:params) do
+      {
+        source_path: 'v1/api/kv/certs/puppet/',
+        vault: true
+      }
+    end
+
+    it {
```
Comment (rnelson0) [inline spec/defines/vhost_spec.rb:72]: We should test some content as well, I’ll look up the exact matches names tomorrow but something like `.with_content(/TEST/)`, just to make sure we are creating the content as expected.

File: manifests/vhost.pp [hunk 2]
```diff
@@ -47,19 +55,38 @@
     fail('You must provide a target_ path for the certs to certs::vhost.')
   }
 
-  $crt = "${name}.crt"
-  $key = "${name}.key"
+  $defaults = { ensure => file, notify => Service[$service] }
```
Comment (rnelson0) [inline manifests/vhost.pp:58]: I don’t see $defaults in use, we could use that and have the conditionals set the additional content, then have one set of resource declarations outside of it. Do you think that would be understandable, or does the duplication make it easier to read?

File: Gemfile [hunk 3]
```diff
@@ -29,6 +29,7 @@ group :development do
   gem "puppet-module-win-default-r#{minor_version}", '~> 0.3',   require: false, platforms: [:mswin, :mingw, :x64_mingw]
   gem "puppet-module-win-dev-r#{minor_version}", '~> 0.3',       require: false, platforms: [:mswin, :mingw, :x64_mingw]
   gem "github_changelog_generator",                              require: false, git: 'https://github.com/skywinder/github-changelog-generator', ref: '20ee04ba1234e9e83eb2ffb5056e23d641c7a018' if Gem::Version.new(RUBY_VERSION.dup) >= Gem::Version.new('2.2.2')
+  gem "rspec-puppet-utils"
```
Comment (rnelson0) [inline Gemfile:32]: This should be added to .sync.yml since this module uses PDK.

**Review by codylane — CHANGES_REQUESTED** [review]
Hi @ekingme, 

I completed my code review and the changes look good.  The only thing I think we are missing here is since we added the `vault_lookup` function our or `certs::vhost` defined type I feel like we should call out that in order to get `vault_lookup` you need to also download `'jsok/vault', '2.3.0'`.  

So I think it's this syntax, it may not be 100% correct. 

```
  "requirements": [
    {
      "name": "puppet",
      "version_requirement": ">= 4.10.0 < 7.0.0"
    },
    {
      "name": "jsok/vault",
      "version_requirement": "2.3.0"
     }
  ],
```

**Comment by codylane** [comment]
> Hi @ekingme,
> 
> I completed my code review and the changes look good. The only thing I think we are missing here is since we added the `vault_lookup` function our or `certs::vhost` defined type I feel like we should call out that in order to get `vault_lookup` you need to also download `'jsok/vault', '2.3.0'`.
> 
> So I think it's this syntax, it may not be 100% correct.
> 
> ```
> {
>   "name": "rnelson0-certs",
>   "version": "1.0.1-rc0",
>   "author": "rnelson0",
>   "summary": "SSL Certificate File Management",
>   "license": "Apache-2.0",
>   "source": "https://github.com/rnelson0/puppet-certs",
>   "project_page": "https://github.com/rnelson0/puppet-certs",
>   "issues_url": "https://github.com/rnelson0/puppet-certs/issues",
>   "dependencies": [
>     {
>       "name": "vsok-vault",
>       "version_requirement": ">= 2.0.3 < 3.0.0"
>     }
>   ],
>   "operatingsystem_support": [
>     {
>       "operatingsystem": "RedHat",
>       "operatingsystemrelease": [
>         "6",
>         "7"
>       ]
>     },
>     {
>       "operatingsystem": "CentOS",
>       "operatingsystemrelease": [
>         "6",
>         "7"
>       ]
>     }
>   ],
>   "requirements": [
>     {
>       "name": "puppet",
>       "version_requirement": ">= 4.10.0 < 7.0.0"
>     }
>   ],
>   "pdk-version": "1.18.1",
>   "template-url": "pdk-default#1.18.1",
>   "template-ref": "tags/1.18.1-0-g3d2e75c"
> }
> ```

Also, to be future complaint with PDK I feel as though we should add the `mock_with: ':rspec'` in our ~/.sync.yml.  Here's the full file for reference.

```
---
Gemfile:
  optional:
    ':development':
      - gem: 'github_changelog_generator'
        git: 'https://github.com/skywinder/github-changelog-generator'
        ref: '20ee04ba1234e9e83eb2ffb5056e23d641c7a018'
        condition: "Gem::Version.new(RUBY_VERSION.dup) >= Gem::Version.new('2.2.2')"
Rakefile:
  default_disabled_lint_checks:
    - '140chars'
    - 'arrow_alignment'
spec/spec_helper.rb:
  hiera_config: 'spec/fixtures/hiera.yaml'
  mock_with: ':rspec'
```

**Comment by rnelson0** [comment]
I'm going to merge this as is, so we can pull from the branch for testing. Before I release a new forge module, let's get the readme updated (rather than add a hard dependency) and add the mock_with.

**Pull request merged**